### PR TITLE
Non-recipe quality selection

### PR DESCRIPTION
**Pull request by DaleStan**
This allows quality selection for everything except recipes, and should reflect the quality-related values everywhere they are relevant.

This corresponds to tasks "Select quality of crafting entity with new GUI" through "Use quality-dependent stats for modules/beacons" and "UnlockQualityModifier in TechnologyPrototype::effects for milestone analysis" in #311.

### CODE REVIEW

File: Yafc.Model/Data/DataClasses.cs [hunk 1]
```diff
@@ -556,18 +695,24 @@ public class EntityEnergy {
...
+    public float baseConsumption { get; internal set; }
+    public float baseSpeed { get; internal set; }
+    public float baseProductivity { get; internal set; }
+    public float basePollution { get; internal set; }
+    public float baseQuality { get; internal set; }
+    public float Consumption(Quality quality) => baseConsumption >= 0 ? baseConsumption : quality.ApplyModuleBonus(baseConsumption);
+    public float Speed(Quality quality) => baseSpeed <= 0 ? baseSpeed : quality.ApplyModuleBonus(baseSpeed);
+    public float Productivity(Quality quality) => baseProductivity <= 0 ? baseProductivity : quality.ApplyModuleBonus(baseProductivity);
+    public float Pollution(Quality quality) => basePollution >= 0 ? basePollution : quality.ApplyModuleBonus(basePollution);
+    public float Quality(Quality quality) => baseQuality <= 0 ? baseQuality : quality.ApplyModuleBonus(baseQuality);
```
Comment (veger) [inline Yafc.Model/Data/DataClasses.cs:715]: Just to be sure:
Some of these use `>=` and other `<=`, it feels (intuitively) wrong. Is this indeed as intended?

File: Yafc.Parser/Data/FactorioDataDeserializer.cs [hunk 2]
```diff
@@ -130,9 +130,13 @@ public Project LoadData(string projectPath, LuaTable data, LuaTable prototypes,
         DeserializePrototypes(raw, "recipe", DeserializeRecipe, progress, errorCollector);
         progress.Report(("Loading", "Loading technologies"));
         DeserializePrototypes(raw, "technology", DeserializeTechnology, progress, errorCollector);
-        progress.Report(("Loading", "Loading entities"));
-        LuaTable entityPrototypes = (LuaTable?)prototypes["entity"] ?? throw new ArgumentException("Could not load prototypes.item from data argument", nameof(prototypes));
+        progress.Report(("Loading", "Loading qualities"));
+        DeserializePrototypes(raw, "quality", DeserializeQuality, progress, errorCollector);
+        Quality.Normal = GetObject<Quality>("normal");
+        rootAccessible.Add(Quality.Normal);
```
Comment (veger) [inline Yafc.Parser/Data/FactorioDataDeserializer.cs:136]: Should this be protected with a check to be sure the `quality` module is loaded/available? (With some default if not?)
So we won't run into issues to support if base game (when the user doesn't have the DLC)

File: changelog.txt [hunk 3]
```diff
@@ -20,6 +20,7 @@ Date:
     Features:
         - Add pseudo-items representing a recipe's total sushi-inputs and sushi-outputs.
           They can be toggled on/off in the LMB of the recipe.
+        - Add the ability to select qualities for crafters, accumulators, modules, and beacons.
```
Comment (veger) [inline changelog.txt:23]: I feel there are some other things fixed as well :wink:

**Review by veger — COMMENTED** [review]
Does this work when a user does not have the DLC (or the quality module disabled)?

File: Yafc.Model/Data/DataClasses.cs (hunk 1 above)
Comment (DaleStan) [reply Yafc.Model/Data/DataClasses.cs:715]: Yes, that's intentional. Beneficial effects are boosted, while detrimental ones are not. The code here checks for detrimental conditions (negative speed, prod, and quality; positive consumption and pollution) and returns the unmodified value in those cases.

File: Yafc.Parser/Data/FactorioDataDeserializer.cs (hunk 2 above)
Comment (DaleStan) [reply Yafc.Parser/Data/FactorioDataDeserializer.cs:136]: The base game defines the qualities normal (in base) and unknown-quality (in core). I've loaded Yafc several times without the quality mod enabled.

It looks like modpacks that don't define a normal quality will have good-enough behavior; a quick check says the quality features in Yafc will be disabled, but everything else will work. (I changed `GetObject<Quality>("normal")` to `GetObject<Quality>("normal2")`. It created a default quality object with no icon and a null `nextQuality`.)

Comment (veger) [reply Yafc.Parser/Data/FactorioDataDeserializer.cs:140]: Ah... so quality is basically build into the base game/mod. And the quality mod will make use of it to extend this feature into something usable?

This seems to make things nice and 'easy' for us. (I was a bit worried that we broke the compatibility with the non-DLC players (as we did a little white back)

File: Yafc.Model/Blueprints/Blueprint.cs [hunk 4]
```diff
@@ -168,3 +169,26 @@ public class BlueprintControlFilter {
...
+}
+
+[Serializable]
+public class BlueprintItemInventory {
+    [JsonPropertyName("in_inventory")] public List<BlueprintInventoryItem> inInventory { get; } = [];
+}
+
+[Serializable]
+public class BlueprintInventoryItem {
+    public int inventory { get; } = 4; // unknown magic number
```
[the remark this answers is not in the document]
Comment (veger) [reply Yafc.Model/Blueprints/Blueprint.cs:192]:
```suggestion
    public int inventory { get; } = 4; // unknown magic number (probably 'modules', needs more investigating)
```

**Review by veger — APPROVED** [review]
The PR is good for me (it is up to you if you agree with my suggestion on not :wink:)

**Pull request merged**